lidar_prod/tasks/utils.py

```diff
@@ -179,36 +179,46 @@ def request_bd_uni_for_building_shapefile(
     bbox: Dict[str, int],
     epsg: int | str,
 ):
-    """Rrequest BD Uni for its buildings.
+    """Request BD Uni for its buildings.
 
-    Create a shapefile with non destructed building on the area of interest
-    and saves it.
+    Create a shapefile with non destructed building on the area of interest and saves it.
 
     Also add a "PRESENCE" column filled with 1 for later use by pdal.
 
+    Note on the projections:
+    Projections are mixed in the BDUni tables.
+    In PostGIS, the declared projection is 0 but the data are stored in the legal projection of the corresponding territories.
+    In each table, there is a a "gcms_territoire" field, which tells the corresponding territory (3 letters code).
+    The gcms_territoire table gives hints on each territory (SRID, footprint)
     """
+
     epsg_srid = (
         epsg if (isinstance(epsg, int) or epsg.isdigit()) else epsg.split(":")[-1]
     )
+
+    sql_territoire = f"""WITH territoire(code) as (SELECT code FROM public.gcms_territoire WHERE srid = {epsg_srid}) """
+
     sql_batiment = f"""SELECT \
         st_setsrid(batiment.geometrie,{epsg_srid}) AS geometry, \
         1 as presence \
-        FROM batiment \
-        WHERE batiment.geometrie \
-        && ST_MakeEnvelope({bbox["x_min"]}, {bbox["y_min"]}, {bbox["x_max"]}, {bbox["y_max"]}, {epsg_srid}) \
+        FROM batiment, territoire \
+        WHERE (batiment.gcms_territoire = territoire.code) \
+        AND batiment.geometrie \
+        && ST_MakeEnvelope({bbox["x_min"]}, {bbox["y_min"]}, {bbox["x_max"]}, {bbox["y_max"]}, 0) \
         AND not gcms_detruit"""
 
     sql_reservoir = f"""SELECT \
         st_setsrid(reservoir.geometrie,{epsg_srid}) AS geometry, \
         1 as presence \
-        FROM reservoir \
-        WHERE reservoir.geometrie \
-        && ST_MakeEnvelope({bbox["x_min"]}, {bbox["y_min"]}, {bbox["x_max"]}, {bbox["y_max"]}, {epsg_srid}) \
+        FROM reservoir, territoire \
+        WHERE (reservoir.gcms_territoire = territoire.code) \
+        AND reservoir.geometrie \
+        && ST_MakeEnvelope({bbox["x_min"]}, {bbox["y_min"]}, {bbox["x_max"]}, {bbox["y_max"]}, 0) \
         AND (reservoir.nature = 'Château d''eau' OR reservoir.nature = 'Réservoir industriel') \
         AND NOT gcms_detruit"""
```

Comment:
@NicolasPyIGN : est-ce que ça te parait pertinent cette mise à jour de la requête ?

Reply:
j'ai l'impression qu'on utilise la géométrie 3d [ici](https://github.com/IGNF/lidar-prod/blob/8824a5b5846d0b0094c758e73b3efc013b42904f/lidar_prod/tasks/building_validation.py#L201) mais je veux bien une confirmation @CharlesGaydon ou @MichelDaab

Reply:
> peut être tester que la bbox tombe dans la bbox de public.gcms_territoire.geometry élargi un peu ?
 @NicolasPyIGN comment ça marche pour tester ça ? la table `public` c'est une table où tu as l'ensemble des territoires existants dans la base de donnée ?

Reply:
> j'ai l'impression qu'on utilise la géométrie 3d [ici](https://github.com/IGNF/lidar-prod/blob/8824a5b5846d0b0094c758e73b3efc013b42904f/lidar_prod/tasks/building_validation.py#L201) mais je veux bien une confirmation @CharlesGaydon ou @MichelDaab

Bâtiment et réservoir sont des MultiPolygone 3D, cf. [Classe Bâtiment sur Pomme](http://pomme.ign.fr/SV3D/Documentation-BDUni-v2/batiment) et [Classe Réservoir sur Pomme](http://pomme.ign.fr/SV3D/Documentation-BDUni-v2/?id_theme=61&id_classe=65).

Je suggérais cette troisième dimension inutile, à gicler via la fonction https://postgis.net/docs/ST_Force2D.html
Je suggérais en outre de conserver une géométrie simplifiée, suffisante à mon sens pour labelliser un quadrant comme conteneur de l'objet, en utilisant https://postgis.net/docs/ST_SimplifyPreserveTopology.html

Reply:
> > peut être tester que la bbox tombe dans la bbox de public.gcms_territoire.geometry élargi un peu ?
> 
> @NicolasPyIGN comment ça marche pour tester ça ? la table `public` c'est une table où tu as l'ensemble des territoires existants dans la base de donnée ?

Le schema Public de la BDUni contient la table des territoires, où chaque enregistrement possède une emprise de validité ; https://pomme.ign.fr/Services-Web/Documentation-Pomme/doku.php?id=bduni:quelques_rappels_pour_manipuler_la_bduni

> Les tables de production (celles du schéma public) mixent les projections. Dans PostGIS, la projection déclarée est 0, mais les données sont bien projetées dans les projections légales selon les territoires. Sur chaque table, on trouve un champ « gcms_territoire » qui indique le territoire (code de trois lettres). La table gcms_territoire permet de donner des indications sur chaque territoire (SRID, emprise). 

Ma suggestion était de dire que la fonction request_bd_uni_for_building_shapefile comprends comme paramètre une bbox et un code epsg, on peut donc dans une approche défensive tester
1. la validité de la bbox en tant que géométrie, en tant que géométrie a qui on affecte le code epsg
2. la validité du code epsg (correspond à un code de type entier à 4 chiffres voire mieux appartenance à une liste dynamiquement créée en requetant les codes epsg de public.gcms_territoire)
3. la cohérence entre ces 2 informations, que la bbox de projection définie (https://postgis.net/docs/ST_SetSRID.html) par l'epsg soit contenue ou intersecte la geometrie public.gcms_territoire.geometry élargie (https://postgis.net/docs/ST_Buffer.html) du territoire correspondant à cet epsg

Reply:
> j'ai l'impression qu'on utilise la géométrie 3d [ici](https://github.com/IGNF/lidar-prod/blob/8824a5b5846d0b0094c758e73b3efc013b42904f/lidar_prod/tasks/building_validation.py#L201) mais je veux bien une confirmation @CharlesGaydon ou @MichelDaab

Nous n'avons besoin que d'un polygone 2D. D'apèrs ce que je comprends des CHANGELOG de pdal, [filters.overlay considère une intersection 2D](https://pdal.io/en/2.6.0/stages/filters.overlay.html), contrairement à [p.ex. filters.crop](https://pdal.io/en/2.6.3/stages/filters.crop.html) qui mentionne explicitement l'usage de la 3D. Donc pas de soucis en l'état.

Et si j'en crois [le commentaire dans le code de création du shapefile après requete de la BD Topo](https://github.com/IGNF/lidar-prod/blob/8824a5b5846d0b0094c758e73b3efc013b42904f/lidar_prod/tasks/utils.py#L217), on lit puis écrit à nouveau le shapefile, ce qui a pour effet apparemment de passer de la 3D à la 2D. Donc pas de risque. On peut éventuellement passer cette étape dans la requete directement (c'est du refactor, donc toujours un petit risque de casser quelque chose...).

Reply:
Si je me souviens bien, la lecture puis réécriture du shapefile a été un workaround car le fichier shapefile qui sortait directement de la requête de la BD UNI provoquait un bug de temps en temps, dans pdal si ma mémoire est bonne.

Comme ce bug n'existe pas avec un fichier écrit via geopandas, on fait cette opération un peu laide mais qu'on n'a par réussi à esquiver autrement

Reply:
> Si je me souviens bien, la lecture puis réécriture du shapefile a été un workaround car le fichier shapefile qui sortait directement de la requête de la BD UNI provoquait un bug de temps en temps, dans pdal si ma mémoire est bonne.
> 
> Comme ce bug n'existe pas avec un fichier écrit via geopandas, on fait cette opération un peu laide mais qu'on n'a par réussi à esquiver autrement

En tant que proche de la majec, ce bug m'interpelle. 

Les géométries BDUni sont forcées pour que les objets soient de srid=0 et surtout de géométrie respectant https://postgis.net/docs/ST_IsValidDetail.html
![image](https://github.com/IGNF/lidar-prod/assets/122606277/a2f08671-2de7-4196-b2d6-5c9ed49dc0a2)

De cette garantie de validité et conformité au standard OGC-SFS (Open Geospatial Consortium - Simple Feature SQL) A MINIMA au moment de l'insertion ou de l'édition des objets (OGC-SFS a subi des révisions au fil des ans), je suis curieux de comprendre ce qui fait râler pdal. Si tu trouves un bout de mémoire, j'en suis preneur.

Reply:
En faisant de l'archéologie sur teams, je suis tombé sur un échange le 27/10/22, où je dis :
_le shapefile qui avait des "line intersections" ? J'ai lancé PDAL sans passer par l'étape du traitement du fichier shapefile, c'est passé._
SI je me souviens, et en me basant sur ce commentaire, des fois les shapefiles en sortie de la requête de BD UNI ont des lignes qui se croisent (line intersection), ce qui faisait planter cette [ligne](https://github.com/IGNF/lidar-prod/blob/8824a5b5846d0b0094c758e73b3efc013b42904f/lidar_prod/tasks/building_validation.py#L201) qui fait passer le shapefile dans pdal, et quand je la désactivais le bug disparaissait

Reply:
J'ai essayé de refaire le bug de mon côté avec un shapefile que j'ai retrouvé de l'époque, sans succès. Tout ce que je peux dire, c'est que des fois, en récupérant le shapefile des bâtiments sur la BD_UNI qui correspond à une dalle lidar, il y avait des points d'intersection qui était indiqués par "line intersections" sur ce [site en ligne de lecture des shapefile](https://mapshaper.org/), et que cela ne passsait pas la ligne de code que j'ai indiquée plus haut (ça restait bloqué, sans bug mais sans aboutir)

évidemment cela arrivait plus souvent en ville, nous on a eu le problème sur Lyon.

Reply:
Considérant que vous recherchez la présence d'un bâtiment dans un quadrant, vous pouvez dans la requête BDuni utiliser https://postgis.net/docs/ST_MakeValid.html avec le paramétrage qui vous sied, ça peut dès la requête source vous enlever d'éventuels problèmes (je suspecte que la contrainte BDUni isvalid soit arrivé en cours de production, et que des géométries anciennes ont pu passer à travers de cette contrainte ; c'est remonté à jean pené qui jettera un oeil à l'occasion).

SELECT st_setsrid(ST_MakeValid(batiment.geometrie ...

Reply:
Jean me confirme "Et par ex. le st_isvalid est lancé sur chaque base de diffusion (et c'est assez normal qu'il ne trouve rien à force)."

De là, je peux penser à :
_ des implémentations de lecture/écriture et validité qui diffèrent un epu ; selon les outils ce n'est pas toujours compatible à 100%
_ les contrôles BDU peut être pas executés sur le backlog d'objets détruits, objets qui pourraient avoir été requêtés dans vos premières expériences si la clause NOT gcms_detruit n'était pas utilisée
_ mystère vaudou

Reply:
Après avoir levé la question en daily, a été évoqué la possiblité que ce fut en effet un bug pdal à l'époque, mais que depuis il a été corrigé.

Mais pour s'en assurer il faudrait retrouver un shapefile avec des "line intersections" et refaire le test